### PR TITLE
Establish pattern system

### PR DESCRIPTION
**Pull request by gruetter**
Hi everyone. I have prepared a document to collect and discuss ideas for a pattern system for the InnerSource patterns in preparation for the Dec. 1st patterns group meeting. It includes a summary of related work on patterns in the software space and a proposal for organising our patterns.

I'd like to invite you to add your own proposals and to provide feedback for other proposals in the PR.

>> Georg

### CODE REVIEW

**Review by rrrutledge — APPROVED** [review]
This looks like a good starting point, especially leading up to Friday's call.  Can anyone else approve also?  I think that one other approval feels like sufficient to merge.  It would be nice to make PRs against this file in _master_.

**Comment by dicortazar** [comment]
I agree that this is a great starting point, thanks for thinking of this problem @gruetter !

However it comes to my mind other characterization and this is based on Jim's talk "InnerSource 101 and The Apache Way"[1]: Culture, Communication, Transparency, Collaboration, Community and Meritocracy as the main points to characterize all of the patterns.

I see the one proposed by @gruetter focused on the process when adopting InnerSource and I like that as well (please correct me if I'm wrong!! U_U ). I would perhaps more generic steps such as: pre-adoption, early adoption, etc...

And then it comes to my mind as third option the way to internally sell InnerSource within an organization and this basically implies all of the departments in that organization that a given pattern can affect. For instance, the 'Review Committee' pattern helps with the process of letting developers work on their own and still give control to middle management and business roles. Would it make sense to have another potential characterization based on the companies structure? I foresee areas such as: DevTeams, Marketing, Management, Legal, Business, Financial, etc...




[1] https://es.slideshare.net/jimjag/apachecon-2017-innersource-and-the-apache-way

**Comment by NewMexicoKid** [comment]
@rrrutledge ,
While I think that the proposal is good, I think that we should leave this PR open for the moment (I added a couple items that need review; not sure if that is the proper process to follow, but it seemed expedient at the time ;-) ).

**Comment by rrrutledge** [comment]
> While I think that the proposal is good, I think that we should leave this PR open for the moment (I added a couple items that need review; not sure if that is the proper process to follow, but it seemed expedient at the time ;-) ).

@NewMexicoKid sounds good.  What are the extra items in the pull request?  For some reason I can't see any extra feedback or comments from you visible to me.

**Comment by rrrutledge** [comment]
> What are the extra items in the pull request

Whoops, nevermind, I see it now @NewMexicoKid

**Comment by gruetter** [comment]
Wow. Excellent contributions, @dicortazar and @NewMexicoKid . Looking forward to tomorrows discussion.

File: pattern-system.md [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+PLoP 2016](http://www.hillside.net/plop/2016/papers/three/26.3.pdf).
+
+## Candiate Classifications
+
+This section shall serve to collect individual proposals for systems of ISC 
+patterns. Contribute away ;)
+
+### Georg Grütter
+
+I propose we use a classification of the problem, a pattern tries to solve as
```
Comment (rrrutledge) [inline pattern-system.md:92]: I think a agree with this sentiment but would like to propose that "the problem" be defined as the intersection of vectors in any number of the planes that we're listing out below.

**Comment by rrrutledge** [comment]
> @rrrutledge ,
> While I think that the proposal is good, I think that we should leave this PR open for the moment (I added a couple items that need review; not sure if that is the proper process to follow, but it seemed expedient at the time ;-) ).

No problem!  My main motivation was to have a place where we could iterate and give more content but it looks like we are just doing that here in this branch!

**Comment by oferher** [comment]
When we first discussed this I thought more about the different activities / implication one needs to address when implementing inner source rather than the adoption stages. But both are very useful and helpful. If we adopt tags, we can have multiple vectors of classifications at once. For example one pattern can be tagged for  #mid-management-buy-in and on the same time #Getting-started-with-InnerSource. Or even one pattern can be applied to different stages of adoption.
A few examples of activities / implication can be: 
- tools
- security
- Culture change (probably need to separate into multiple categories)
   - buy in
   - trust
   - etc.
- Processes / Methodology
- Coding practices

**Review by rshanmer — COMMENTED** [review]
The classifications that resonated most with me are Georg's (Getting started, middle mgmt, dev buy-in, fostering contributions, and interfacing).  at my first thought these are helpful as I'm starting a program of inner sourcing.  

However -- who's our intended audience for the patterns?  is it someone just getting started or people in the thick of doing things that want extra help with tools or product development or innovation (e.g. Tim's list)?  

... I still like Georg's.

To add to what Tim wrote -- the GOF book is a collection of patterns.  I agree that we want to create a language (it's cooler and ultimately more useful).  But my experience suggests that we're better off building a catalog first, then reflecting on the catalog and seeing what gaps there are in a language that we need to fill in and then address them to make a language.  My $.02 is that we should keep the idea of a language in the back of our minds and just try to cover the terrain with a rough map that we can fill in between.  When do we start filling in?  Once we are pretty confident that we have basic topography covered (or started) in each of our classification areas.

**Review by NewMexicoKid — COMMENTED** [review]
Looks okay to be one of our published documents; went ahead and merged this. We can later edit this and/or augment it with other documents.

**Comment by rrrutledge** [comment]
Woah, for something so old I think it would have been good to wait a bit and bring up the topic again with folks.  Was something else dependent on this pull getting merged?

**Comment by NewMexicoKid** [comment]
No, no dependencies. Just thought it was safe to merge (and reduce our open PR count). 
It's just a collection of different categorization schemes. We can later edit it with a new PR when there are further thoughts on categories/pattern language.

**Pull request merged**